### ISSUE TITLE
1

### ISSUE THREAD

**Issue by Stick-To** [Stick-To/fc_net-with-matlab#2]
`import numpy as np

def batchnorm_forward(x, gamma, beta, bn_param):
    mode = bn_param['mode']
    eps = bn_param.get('eps', 1e-5)
    momentum = bn_param.get('momentum', 0.9)
    N,D= x.shape
    running_mean = bn_param.get('running_mean', np.zeros(D, dtype=x.dtype))
    running_var = bn_param.get('running_var', np.zeros(D, dtype=x.dtype))
    if mode == 'train':
        sample_mean=np.mean(x,axis=0)
        sample_var=np.var(x,axis=0)
        x_=(x-sample_mean)/(sample_var+eps)**0.5
        out=gamma*x_+beta
        running_mean = momentum * running_mean + (1 - momentum) * sample_mean
        running_var = momentum * running_var + (1 - momentum) * sample_var
        cache=(x,x_,gamma,beta,sample_mean,sample_var,eps)
    else:
        out=(x-bn_param['running_mean'])*gamma/(bn_param['running_var']+eps)**0.5+beta
    bn_param['running_mean'] = running_mean
    bn_param['running_var'] = running_var
    return out, cache

def batchnorm_backward(dout, cache):
    # dout: 上游梯度
    N=dout.shape[0]
    x,x_,gamma,beta,sample_mean,sample_var,eps=cache
    dx_=dout*gamma
    dsample_var=np.sum(dx_*(x-sample_mean),axis=0)*(-0.5)*(sample_var+eps)**(-1.5)
    dsample_mean=np.sum(-dx_/(sample_var+eps)**0.5,axis=0)+dsample_var*np.sum(-2*(x-sample_mean),axis=0)/N
    dx=dx_/(sample_var+eps)**0.5+dsample_var*2*(x-sample_mean)/N+dsample_mean/N
    dgamma=np.sum(dout*x_,axis=0)
    dbeta=np.sum(dout,axis=0)
    return dx, dgamma, dbeta

def dropout_forward(x, dropout_param):
    p, mode = dropout_param['p'], dropout_param['mode']
    if 'seed' in dropout_param:
        np.random.seed(dropout_param['seed'])
    if mode == 'train':
        mask=(np.random.rand(*x.shape)<p)/p
        out=x*mask
    elif mode == 'test':  
        out=x
    cache = (dropout_param, mask)
    out = out.astype(x.dtype, copy=False)
    return out, cache

def dropout_backward(dout, cache):
    # dout: 上游梯度
    dropout_param, mask = cache
    mode = dropout_param['mode']
    if mode == 'train':
        dx=dout*mask
    elif mode == 'test':
        dx = dout
    return dx

`
